### PR TITLE
Harmonizing textile and food simulators

### PR DESCRIPTION
**Pull request by magopian**
[user story](https://www.notion.so/Brainstorming-vue-d-ensemble-volutions-front-textile-9c72a35830d74077bb950c6e5e49a75c)

### CODE REVIEW

File: src/Views/BaseComponent.elm
Comment (magopian) [inline src/Views/BaseComponent.elm:19]: Hopefully it's clear enough that here the `BaseComponent` represents either
- a `MaterialInput`
- a `RecipeIngredient`

In both cases there is a component (`Material` or `Ingredient`), and a quantity (`Mass` or `Split`) and a country.

File: src/Views/Component/SplitInput.elm
Comment (magopian) [inline src/Views/Component/SplitInput.elm:1]: This is a bit like `MassInput`, but for `MaterialInput.share`

**Review by magopian — COMMENTED** [review]
This PR groups a few changes that are interconnected:
- no more tabs for textile: no more "detailed view" nor "dataviz" tabs
- use a generic "BaseComponent" to display and update base components, their quantity and their country
- use a modal to select base components

File: src/Views/BaseComponent.elm
Comment (n1k0) [inline src/Views/BaseComponent.elm:43]: How about ordering these alphabetically?

File: src/Views/BaseComponent.elm
Comment (n1k0) [inline src/Views/BaseComponent.elm:60]: Nit: how about:

```elm
        autocompleteState =
            db.components
                |> List.filter (\component -> not (List.member component excluded))
                |> AutocompleteSelector.init toString
```

File: src/Views/Textile/Step.elm
Comment (n1k0) [inline src/Views/Textile/Step.elm:47]: Nit: I'd really like these properties to be ordered alphabetically

File: src/Views/Textile/Step.elm
Comment (n1k0) [inline src/Views/Textile/Step.elm:356]: I think you meant `modal`

File: src/Views/Textile/Step.elm
Comment (n1k0) [inline src/Views/Textile/Step.elm:708]: Nit: This `let…in` statement is huge, could we extract part of it into external functions, mostly for legibility purpose?

File: src/Views/Textile/Step.elm
Comment (n1k0) [inline src/Views/Textile/Step.elm:760]: Nit: would gain in legibility spread over multiple lines

File: styles.scss
Comment (n1k0) [inline styles.scss:842]: Nit: the multiple spaces separators were there to align the blocks, which isn't applicable anymore

**Review by n1k0 — CHANGES_REQUESTED** [review]
Nice work! The patch is huge and I'm pretty sure I missed a few things in my review. 

My main concern to address before landing is that we lost a nice feature which allowed to navigate autocomplete results using the `Tab` and arrow keys, we should add it back. This may just be matter of ensuring the autocomplete search input gains focus when the modal opens and we start typing, which is not the case anymore with this patch.

Otherwide and again, great work 👍

File: index.js
Comment (n1k0) [reply index.js:99]: Now I think about it, it was for automatically selecting the url to share when clicking on the text input. But then we added the button for copying the link to the clipboard, so it's not useful anymore. Good thing you removed it!

File: src/Views/BaseComponent.elm
Comment (magopian) [reply src/Views/BaseComponent.elm:19]: I quite like "Element", I'll give it a go and see how it unfolds, thanks for the suggestions.

File: src/Views/Textile/Step.elm
Comment (magopian) [reply src/Views/Textile/Step.elm:833]: I changed that to either display "Ajouter une matière"

Capture d’écran 2023-10-02 à 10 52 58

or:

Capture d’écran 2023-10-02 à 10 53 40

File: src/Page/Textile/Simulator/ViewMode.elm
Comment (magopian) [reply src/Page/Textile/Simulator/ViewMode.elm:9]: Good point

Comment (magopian) [reply src/Page/Textile/Simulator/ViewMode.elm:9]: We might also get rid of the "detailed/simple" difference to harmonize with the food simulator, at some point... to be discussed ;)

File: src/Views/AutocompleteSelector.elm
Comment (magopian) [reply src/Views/AutocompleteSelector.elm:48]: Good catch, thanks

File: src/Views/Textile/Step.elm
Comment (magopian) [reply src/Views/Textile/Step.elm:704]: Extracted most of it in a `createConfig` helper 👍

**Comment by magopian** [comment]
@n1k0 I believe i've addressed your comments, let me know what you think?

File: src/Page/Food.elm
Comment (n1k0) [inline src/Page/Food.elm:547]: Nit: I realize this now lives at the root of `Page/Food.elm`, so maybe we should name it more explicitely eg. `createElementSelectorConfig`

File: src/Page/Textile/Simulator.elm
Comment (n1k0) [inline src/Page/Textile/Simulator.elm:498]: Putting this here but it's probably not in this specific code block that it should be done: when testing the UI, I realize when selecting an Element using the keyboard (arrow keys then pressing Enter), the modal closes and the focus is then positioned on the "add bookmark button" in the sidebar, which makes using the app using the keyboard difficult (you have to switch to using the mouse).

What would be great is to set the focus on the quantity input after an element is selected and the autocomplete modal closes, so we can immediately define it's 50% or 200g, then tab again to gain focus on the "add new element" btn, rinse and repeat. Full keyboard navigation.

I have a feeling it's easy to implement in the update function when an element is selected but I may be wrong; if it's trivial, let's impolement it in this PR, if it's not, let's create a card because I think this would add significant UX value to the app.

File: src/Route.elm
Comment (n1k0) [inline src/Route.elm:84]: Nit: If we want to keep this BC bit, maybe extract this in some `deprecatedTextileRouteParser` fn to easily identify and remove it when the time comes?

I'd go with removing this BC layer entirely, but I'll leave the final decision up to you.

File: src/Views/BaseElement.elm
Comment (n1k0) [inline src/Views/BaseElement.elm:119]: Nit: how about typing this `Bool` explicitly as `{ disabled : Bool }`, so it can be called as `deleteItemButton { disabled = List.length excluded == 1 } deleteEvent`?

**Review by n1k0 — APPROVED** [review]
Amazing work! A few bits you want to address or not, in this patch or in a next PR.

File: src/Route.elm
Comment (magopian) [reply src/Route.elm:84]: Good point on extracting it. I believe it doesn't cost much to keep it for now, this way we won't brake bookmarks. Maybe we'll revisit (and remove) in the future?

**Comment by magopian** [comment]
Ok I've addressed your comments... but the "manage the focus". I'm working on a solution to that, but sadly it seems to be a bit complex. I'll have a solution to test shortly, see if it works well enough, and warrants the added code complexity.

**Comment by n1k0** [comment]
@magopian yeah this kind of stuff can get tricky, don't bother if it's too complicated, we can address that later

**Comment by magopian** [comment]
@n1k0 what do you think about the latest commit? Is it working for you?

File: src/Views/BaseElement.elm
Comment (n1k0) [inline src/Views/BaseElement.elm:136]: Maybe use the id, I'm slightly uncomfortable with any possible character being used as DOM ids

image

**Comment by n1k0** [comment]
> @n1k0 what do you think about the latest commit? Is it working for you?

@magopian it now focuses on the selector itself rather than the quanbtity input but that works for me 👍

**Comment by magopian** [comment]
Updated

**Comment by n1k0** [comment]
@magopian ⛴️

**Pull request merged**
